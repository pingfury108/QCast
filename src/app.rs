use async_trait::async_trait;
use loco_rs::{
    app::{AppContext, Hooks, Initializer},
    bgworker::Queue,
    boot::{create_app, BootResult, StartMode},
    config::Config,
    controller::AppRoutes,
    db::{self, truncate_table},
    environment::Environment,
    task::Tasks,
    Result,
};
use migration::Migrator;
use std::path::Path;

#[allow(unused_imports)]
use crate::{controllers, initializers, models::_entities::users, tasks, workers};

pub struct App;
#[async_trait]
impl Hooks for App {
    fn app_name() -> &'static str {
        env!("CARGO_CRATE_NAME")
    }

    fn app_version() -> String {
        format!(
            "{} ({})",
            env!("CARGO_PKG_VERSION"),
            option_env!("BUILD_SHA")
                .or(option_env!("GITHUB_SHA"))
                .unwrap_or("dev")
        )
    }

    async fn boot(
        mode: StartMode,
        environment: &Environment,
        config: Config,
    ) -> Result<BootResult> {
        create_app::<Self, Migrator>(mode, environment, config).await
    }

    async fn initializers(_ctx: &AppContext) -> Result<Vec<Box<dyn Initializer>>> {
        Ok(vec![])
    }

    fn routes(_ctx: &AppContext) -> AppRoutes {
        AppRoutes::with_default_routes() // controller routes below
            .add_route(controllers::books::routes())
            .add_route(controllers::chapters::routes())
            .add_route(controllers::medias::routes())
            .add_route(controllers::public::routes())
            .add_route(controllers::auth::routes())
            .add_route(controllers::dashboard::routes())
            // 后台管理路由
            .add_route(controllers::admin::users::routes())
            .add_route(controllers::admin::groups::routes())
    }
    async fn connect_workers(_ctx: &AppContext, _queue: &Queue) -> Result<()> {
        // 暂无 worker
        Ok(())
    }

    #[allow(unused_variables)]
    fn register_tasks(tasks: &mut Tasks) {
        tasks.register(tasks::create_superadmin::CreateSuperadmin);
        // tasks-inject (do not remove)
    }
    async fn truncate(ctx: &AppContext) -> Result<()> {
        truncate_table(&ctx.db, users::Entity).await?;
        Ok(())
    }
    async fn seed(ctx: &AppContext, base: &Path) -> Result<()> {
        db::seed::<users::ActiveModel>(&ctx.db, &base.join("users.yaml").display().to_string())
            .await?;
        Ok(())
    }
}
