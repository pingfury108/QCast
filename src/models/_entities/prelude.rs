//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

pub use super::books::Entity as Books;
pub use super::chapters::Entity as Chapters;
pub use super::medias::Entity as Medias;
pub use super::site_settings::Entity as SiteSettings;
pub use super::user_group_members::Entity as UserGroupMembers;
pub use super::user_groups::Entity as UserGroups;
pub use super::users::Entity as Users;
