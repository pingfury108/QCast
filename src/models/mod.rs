pub mod _entities;
pub mod books;
pub mod chapters;
pub mod medias;
pub mod user_group_members;
pub mod user_groups;
pub mod users;
