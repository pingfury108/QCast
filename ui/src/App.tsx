import { QueryClientProvider } from '@tanstack/react-query';
import { ReactQueryDevtools } from '@tanstack/react-query-devtools';
import { RouterProvider } from 'react-router-dom';
import { Toaster } from 'sonner';
import { queryClient } from './lib/query-client';
import { router } from './routes';

function App() {
  return (
    <QueryClientProvider client={queryClient}>
      <RouterProvider router={router} />
      <Toaster position="top-right" richColors />
      <ReactQueryDevtools initialIsOpen={false} />
    </QueryClientProvider>
  );
}

export default App;
