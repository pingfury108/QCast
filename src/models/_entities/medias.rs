//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "medias")]
pub struct Model {
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
    #[sea_orm(primary_key)]
    pub id: i32,
    pub title: String,
    #[sea_orm(column_type = "Text", nullable)]
    pub description: Option<String>,
    pub file_type: String,
    pub file_path: String,
    pub file_size: Option<i64>,
    pub duration: Option<i32>,
    pub mime_type: Option<String>,
    pub access_token: String,
    pub access_url: Option<String>,
    pub qr_code_path: Option<String>,
    pub file_version: i32,
    pub original_filename: Option<String>,
    pub play_count: i32,
    pub is_public: bool,
    pub chapter_id: Option<i32>,
    pub book_id: i32,
    pub user_id: i32,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}
