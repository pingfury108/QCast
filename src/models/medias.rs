pub use super::_entities::books;
pub use super::_entities::medias::{ActiveModel, Column, Entity, Model};
pub use super::_entities::users;
use sea_orm::entity::prelude::*;
use sea_orm::QueryOrder;
use uuid::Uuid;
pub type Medias = Entity;

#[async_trait::async_trait]
impl ActiveModelBehavior for ActiveModel {
    async fn before_save<C>(self, _db: &C, insert: bool) -> std::result::Result<Self, DbErr>
    where
        C: ConnectionTrait,
    {
        let mut this = self;

        if insert {
            // 设置默认值
            if this.file_version.is_unchanged() {
                this.file_version = sea_orm::ActiveValue::Set(Some(1));
            }
            if this.play_count.is_unchanged() {
                this.play_count = sea_orm::ActiveValue::Set(Some(0));
            }
            if this.is_public.is_unchanged() {
                this.is_public = sea_orm::ActiveValue::Set(Some(false));
            }
            // 生成访问令牌
            if this.access_token.is_unchanged() {
                this.access_token = sea_orm::ActiveValue::Set(Some(Uuid::new_v4().to_string()));
            }
        }

        if !insert && this.updated_at.is_unchanged() {
            this.updated_at = sea_orm::ActiveValue::Set(chrono::Utc::now().into());
        }

        Ok(this)
    }
}

// implement your read-oriented logic here
impl Model {
    /// 获取用户的所有媒体文件
    pub async fn find_by_user(db: &DatabaseConnection, user_id: i32) -> Result<Vec<Model>, DbErr> {
        Entity::find()
            .filter(Column::UserId.eq(user_id))
            .order_by_desc(Column::CreatedAt)
            .all(db)
            .await
    }

    /// 获取书籍的所有媒体文件
    pub async fn find_by_book(db: &DatabaseConnection, book_id: i32) -> Result<Vec<Model>, DbErr> {
        Entity::find()
            .filter(Column::BookId.eq(book_id))
            .order_by_asc(Column::CreatedAt)
            .all(db)
            .await
    }

    /// 获取章节的所有媒体文件
    pub async fn find_by_chapter(
        db: &DatabaseConnection,
        chapter_id: i32,
    ) -> Result<Vec<Model>, DbErr> {
        Entity::find()
            .filter(Column::ChapterId.eq(chapter_id))
            .order_by_asc(Column::CreatedAt)
            .all(db)
            .await
    }

    /// 搜索媒体文件
    pub async fn search(
        db: &DatabaseConnection,
        user_id: i32,
        query: &str,
    ) -> Result<Vec<Model>, DbErr> {
        Entity::find()
            .filter(Column::UserId.eq(user_id))
            .filter(
                Column::Title
                    .contains(query)
                    .or(Column::Description.contains(query))
                    .or(Column::OriginalFilename.contains(query)),
            )
            .order_by_desc(Column::CreatedAt)
            .all(db)
            .await
    }

    /// 根据访问令牌查找媒体文件
    pub async fn find_by_access_token(
        db: &DatabaseConnection,
        access_token: &str,
    ) -> Result<Option<Model>, DbErr> {
        Entity::find()
            .filter(Column::AccessToken.eq(access_token))
            .one(db)
            .await
    }

    /// 获取公开的媒体文件
    pub async fn find_public(db: &DatabaseConnection) -> Result<Vec<Model>, DbErr> {
        Entity::find()
            .filter(Column::IsPublic.eq(true))
            .order_by_desc(Column::CreatedAt)
            .all(db)
            .await
    }

    /// 增加播放次数
    pub async fn increment_play_count(db: &DatabaseConnection, media_id: i32) -> Result<(), DbErr> {
        Entity::update_many()
            .filter(Column::Id.eq(media_id))
            .col_expr(Column::PlayCount, Expr::col(Column::PlayCount).add(1))
            .exec(db)
            .await?;
        Ok(())
    }
}

// implement your write-oriented logic here
impl ActiveModel {}

// implement your custom finders, selectors oriented logic here
impl Entity {}
