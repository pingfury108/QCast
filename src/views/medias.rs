use serde::{Deserialize, Serialize};
use crate::models::_entities::medias::Model;

#[derive(Debug, Serialize, Deserialize)]
pub struct MediaResponse {
    pub id: i32,
    pub title: Option<String>,
    pub description: Option<String>,
    pub file_type: Option<String>,
    pub file_path: Option<String>,
    pub file_size: Option<i64>,
    pub duration: Option<i32>,
    pub mime_type: Option<String>,
    pub access_token: Option<String>,
    pub access_url: Option<String>,
    pub qr_code_path: Option<String>,
    pub file_version: Option<i32>,
    pub original_filename: Option<String>,
    pub play_count: Option<i64>,
    pub is_public: Option<bool>,
    pub chapter_id: Option<i32>,
    pub book_id: i32,
    pub user_id: i32,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

impl From<Model> for MediaResponse {
    fn from(media: Model) -> Self {
        Self {
            id: media.id,
            title: media.title,
            description: media.description,
            file_type: media.file_type,
            file_path: media.file_path,
            file_size: media.file_size,
            duration: media.duration,
            mime_type: media.mime_type,
            access_token: media.access_token,
            access_url: media.access_url,
            qr_code_path: media.qr_code_path,
            file_version: media.file_version,
            original_filename: media.original_filename,
            play_count: media.play_count,
            is_public: media.is_public,
            chapter_id: media.chapter_id,
            book_id: media.book_id,
            user_id: media.user_id,
            created_at: media.created_at.into(),
            updated_at: media.updated_at.into(),
        }
    }
}

/// 公开访问的媒体响应（不包含敏感信息）
#[derive(Debug, Serialize, Deserialize)]
pub struct PublicMediaResponse {
    pub id: i32,
    pub title: Option<String>,
    pub description: Option<String>,
    pub file_type: Option<String>,
    pub duration: Option<i32>,
    pub mime_type: Option<String>,
    pub original_filename: Option<String>,
    pub play_count: Option<i64>,
    pub created_at: chrono::DateTime<chrono::Utc>,
}

impl From<Model> for PublicMediaResponse {
    fn from(media: Model) -> Self {
        Self {
            id: media.id,
            title: media.title,
            description: media.description,
            file_type: media.file_type,
            duration: media.duration,
            mime_type: media.mime_type,
            original_filename: media.original_filename,
            play_count: media.play_count,
            created_at: media.created_at.into(),
        }
    }
}