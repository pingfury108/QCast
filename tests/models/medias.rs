use insta::{assert_debug_snapshot, with_settings};
use loco_rs::testing::prelude::*;
use qcast::app::App;
use qcast::models::_entities::medias::ActiveModel;
use qcast::models::medias;
use sea_orm::{ActiveModelTrait, Set};
use serial_test::serial;

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_search_medias_by_book() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建两个测试书籍
    let book1 = qcast::models::_entities::books::ActiveModel {
        title: Set("书籍1".to_string()),
        user_id: Set(1),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    let book2 = qcast::models::_entities::books::ActiveModel {
        title: Set("书籍2".to_string()),
        user_id: Set(1),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 在书籍1中创建媒体
    let media1 = ActiveModel {
        title: Set("目标媒体1".to_string()),
        description: Set(Some("这是第一个目标媒体".to_string())),
        book_id: Set(book1.id),
        user_id: Set(1),
        file_path: Set("/test/media1.mp3".to_string()),
        file_type: Set("audio".to_string()),
        file_size: Set(Some(1024)),
        original_filename: Set(Some("media1.mp3".to_string())),
        access_token: Set("test_token_1".to_string()),
        file_version: Set(1),
        play_count: Set(0),
        is_public: Set(false),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    let media2 = ActiveModel {
        title: Set("普通媒体".to_string()),
        description: Set(Some("这是一个普通媒体".to_string())),
        book_id: Set(book1.id),
        user_id: Set(1),
        file_path: Set("/test/media2.mp3".to_string()),
        file_type: Set("audio".to_string()),
        file_size: Set(Some(2048)),
        original_filename: Set(Some("media2.mp3".to_string())),
        access_token: Set("test_token_2".to_string()),
        file_version: Set(1),
        play_count: Set(0),
        is_public: Set(false),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    let media3 = ActiveModel {
        title: Set("目标媒体2".to_string()),
        description: Set(Some("这是第二个目标媒体".to_string())),
        book_id: Set(book1.id),
        user_id: Set(1),
        file_path: Set("/test/media3.mp3".to_string()),
        file_type: Set("audio".to_string()),
        file_size: Set(Some(3072)),
        original_filename: Set(Some("media3.mp3".to_string())),
        access_token: Set("test_token_3".to_string()),
        file_version: Set(1),
        play_count: Set(0),
        is_public: Set(false),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 在书籍2中创建一个包含"目标"关键词的媒体
    let _media4 = ActiveModel {
        title: Set("书籍2的目标媒体".to_string()),
        description: Set(Some("这是书籍2的媒体".to_string())),
        book_id: Set(book2.id),
        user_id: Set(1),
        file_path: Set("/test/media4.mp3".to_string()),
        file_type: Set("audio".to_string()),
        file_size: Set(Some(4096)),
        original_filename: Set(Some("media4.mp3".to_string())),
        access_token: Set("test_token_4".to_string()),
        file_version: Set(1),
        play_count: Set(0),
        is_public: Set(false),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 使用 search_by_book 搜索书籍1中包含"目标"的媒体
    let search_results = medias::Model::search_by_book(&boot.app_context.db, 1, book1.id, "目标")
        .await
        .unwrap();

    // 应该返回书籍1中的2个目标媒体，不包含书籍2的媒体
    assert_eq!(search_results.len(), 2);

    // 验证只包含书籍1的媒体
    let ids: Vec<i32> = search_results.iter().map(|m| m.id).collect();
    assert!(ids.contains(&media1.id));
    assert!(ids.contains(&media3.id));
    assert!(!ids.contains(&media2.id)); // 普通媒体不应该在结果中
    assert!(!ids.contains(&_media4.id)); // 书籍2的媒体不应该在结果中

    with_settings!({
        filters => cleanup_media_model()
    }, {
        assert_debug_snapshot!(search_results);
    });
}

fn cleanup_media_model() -> Vec<(&'static str, &'static str)> {
    vec![
        (r"id: \d+,", "id: ID"),
        (r"book_id: \d+,", "book_id: BOOK_ID"),
        (r"user_id: \d+,", "user_id: USER_ID"),
        (r"chapter_id: Some\(\d+\)", "chapter_id: Some(CHAPTER_ID)"),
        (
            r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?\+\d{2}:\d{2}",
            "DATE",
        ),
    ]
}
