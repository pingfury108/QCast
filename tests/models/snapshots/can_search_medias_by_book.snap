---
source: tests/models/medias.rs
expression: search_results
---
[
    Model {
        created_at: DATE,
        updated_at: DATE,
        id: ID
        title: "目标媒体1",
        description: Some(
            "这是第一个目标媒体",
        ),
        file_type: "audio",
        file_path: "/test/media1.mp3",
        file_size: Some(
            1024,
        ),
        duration: None,
        mime_type: None,
        access_token: "test_token_1",
        access_url: None,
        qr_code_path: None,
        file_version: 1,
        original_filename: Some(
            "media1.mp3",
        ),
        play_count: 0,
        is_public: false,
        chapter_id: None,
        book_id: ID
        user_id: ID
    },
    Model {
        created_at: DATE,
        updated_at: DATE,
        id: ID
        title: "目标媒体2",
        description: Some(
            "这是第二个目标媒体",
        ),
        file_type: "audio",
        file_path: "/test/media3.mp3",
        file_size: Some(
            3072,
        ),
        duration: None,
        mime_type: None,
        access_token: "test_token_3",
        access_url: None,
        qr_code_path: None,
        file_version: 1,
        original_filename: Some(
            "media3.mp3",
        ),
        play_count: 0,
        is_public: false,
        chapter_id: None,
        book_id: ID
        user_id: ID
    },
]
